### DISCUSSION TITLE
cookies() does not set values in browser properly

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#55370] [App Router] [answered]
### issue
I want to verify that a cookie is saved in the browser using [cookies](https://nextjs.org/docs/app/api-reference/functions/cookies#cookiesgetall). However, I can't get the browser to save the cookie by doing something like `cookies().set('hoge', 'lee');`.

## I did 

1. call api/route.ts from layout file
2. set cookie in route.ts
3. cookie is set in the log but not saved in the browser

<img width="458" alt="スクリーンショット 2023-09-14 19 16 47" src="https://github.com/vercel/next.js/assets/138372681/0e57b2b8-f7c4-4040-b575-5d7a544f5903">

---

### sample
- src/app/layout.tsx
```tsx
export default async function RootLayout({ children }: { children: React.ReactNode }) {
  const response = await fetch('http://localhost:3000/api').then((res) => res.json());
  console.log(response); // { message: 'set cookie!!' }
  return (
    <html lang="en">
      <InitialfetchWrapper>
        <body className={inter.className}>{children}</body>
      </InitialfetchWrapper>
    </html>
  );
}
```

- src/app/layout.tsx
```tsx
export async function GET() {
  console.log('cookies().getAll() is ', cookies().getAll()); // log => []
  cookies().set('hoge', 'lee');
  cookies().set('fuga', 'lee', { secure: true });
  cookies().set({
    name: 'piyo',
    value: 'lee',
    httpOnly: true,
    path: '/',
  });
  console.log('cookies().getAll() is ', cookies().getAll());
  // log =>
  // [
  //   { name: 'hoge', value: 'lee', path: '/' },
  //   { name: 'fuga', value: 'lee', secure: true, path: '/' },
  //   { name: 'piyo', value: 'lee', httpOnly: true, path: '/' },
  // ];

  return NextResponse.json({ message: 'set cookie!!' });
}

```

### supplement

However, when I made a request directly to http://localhost:3000/api in my browser, a cookie was set. However, this is not the expected behavior.

<img width="1409" alt="スクリーンショット 2023-09-14 19 13 57" src="https://github.com/vercel/next.js/assets/138372681/22c36c2a-791e-46f7-a8db-1652142181c4">

**Comment by icyJoseph** [comment]
Hi,

According to the docs:

- https://nextjs.org/docs/app/building-your-application/routing/route-handlers#cookies

You should do it like this:

```ts
import { cookies } from 'next/headers'
 
export async function GET(request: Request) {
  const cookieStore = cookies()
  const token = cookieStore.get('token')
 
  return new Response('Hello, Next.js!', {
    status: 200,
    headers: { 'Set-Cookie': `token=${token.value}` },
  })
}
```

Notice they include it in the response object.

**Reply by ghost** [reply]
@icyJoseph 

Thanks. However, when I make a request to http://localhost:3000/ and call the api/route.ts file from layout.ts, the cookie does not seem to be saved in the browser.


src/app/api/route.ts

```tsx
export async function GET() {
  // I want to get the csrf token from the api server here and set it in the cookie.
  // const csrfToken = await myFetch('http://localhost:3005/auth/csrf', {
  //   cache: 'no-cache',
  //   credentials: 'include',
  // });
  const csrfToken = 'csrfToken_csrfToken_csrfToken';
  return new Response('Hello, Next.js!', {
    status: 200,
    headers: { 'Set-Cookie': `csrf-token=${csrfToken}` },
  });
}

```


src/app/layout.tsx
```tsx
export default async function RootLayout({ children }: { children: React.ReactNode }) {
  const response = await fetch('http://localhost:3000/api', {
    cache: 'no-store',
    credentials: 'include',
  }).then((res) => res);
  console.log(response.statusText); // OK

  return (
    <html lang="en">
      {/* not use */}
      <InitialfetchWrapper>
        <body className={inter.className}>{children}</body>
      </InitialfetchWrapper>
    </html>
  );
}
```

**Reply by icyJoseph** [reply]
Nah that's not gonna fly. Why would you expect it to work?  Layout is itself preparing a different response entirely. 

Also, you are already on the server, while in layout, so you don't have to make an API request, which btw might fail when building. Just do the work you need right there on layout. That's why we have server components.

**Reply by icyJoseph** [reply]
Also the docs say:

> Good to know: HTTP does not allow setting cookies after streaming starts, so you must use .set() in a Server Action or Route Handler.

I guess the solution here is to use middleware? https://nextjs.org/docs/app/building-your-application/routing/middleware#using-cookies

**Reply by ghost** [reply]
> Nah that's not gonna fly. Why would you expect it to work? Layout is itself preparing a different response entirely.

When I thought about where to set the csrf-token for pre-processing, I thought I could call the Route Handler from layout.ts, but I guess I didn't understand the `HTTP does not allow setting cookies after streaming starts ` part.I will check the documentation again.

> I guess the solution here is to use middleware? https://nextjs.org/docs/app/building-your-application/routing/middleware#using-cookies

Thanks. I was able to solve the problem by using middleware as you advised.

**Reply by icyJoseph** [reply, marked as answer]
Cool! :) Please don't forget to mark as answered. Also feel free to ask further.

**Reply by ghost** [reply]
Thank you kindly!
